Bug 11471448 - Bring tab swipe deletion to designed interaction

sleroux opened the pull request:
* Added panning gesture on cells in the TabTrayController to allow users
  to swipe away their tabs to delete
* Renamed CustomCell to TabCell

st3fan commented:
This is wonderful. Do you think we can apply the same technique to a `UICollectionView`? Asking because @darrinhenein has a patch in progress that replaces the current `UITableView` with a `UICollectionView`.

I'll add some comments to the PR anyway.

st3fan commented on Client/Frontend/Browser/TabTrayController.swift:
```diff
@@ -17,19 +17,39 @@ private struct TabTrayControllerUX {
+    static let totalScale = CGFloat(0.9)
+    static let totalAlpha = CGFloat(0.7)
+    static let minExitVelocity = 800.0
+    static let recenterAnimationDuration = 0.15
```
Can you make all these start with an uppercase letter. That is what we do in the other `*UX` structs too, to note that these things are constants.

st3fan commented on Client/Frontend/Browser/TabTrayController.swift:
```diff
@@ -117,35 +174,94 @@ private class CustomCell: UITableViewCell {
+        switch (recognizer.state) {
+        case .Began:
+            self.startLocation = self.backgroundHolder.center;
+            break
```
Nit: You don't need the `break` in Swift if there is at least one line of code in the `case`.

st3fan commented on Client/Frontend/Browser/TabTrayController.swift:
```diff
@@ -117,35 +174,94 @@ private class CustomCell: UITableViewCell {
+                    self.backgroundHolder.alpha = 1
+                })
+            }
+            break
```
Is it possible to do as much as possible using the `CGFloat` type in the above code for `.Changed` and `.Ended` ? There is a lot of casting between `CGFloat`, `Double` and `Int` going on and I think most of that can go away.

sleroux commented:
@st3fan @ncalexan I extracted the animation code into it's own SwipeAnimator object that takes in any child/parent view relationship and performs the swipe. The animation code will work when this moves over to a UICollectionView as well. Also cleaned up the casting in the panning selector code.

st3fan commented:
@darrinhenein @wesj what do you think of merging this in already? i think people will appreciate this change. i think we can do it before we switch to a `UICollectionView`.

darrinhenein commented:
@st3fan works for me. Seems to be nicely refactored for our eventual port to UICollectionView as well.

darrinhenein commented:
@sleroux does this still have the offset issue on 6/6+?

st3fan commented:
@darrinhenein it does. but lets fix that as a followup. i'd love to get this in the Aurora build I'm sending out today to get feedback.

darrinhenein commented:
:shipit: Ship it.

st3fan commented:
@sleroux I moved this branch to https://github.com/mozilla/firefox-ios/pull/293 so that I could add an additional change to it. I did two things:

* To fix the offset issue I assigned `self.startLocation` to `self.animatingView.center` instead of `self.originalCenter`.
* I changed the logic to decide if the tab should bounce back or not a little bit. Previously it was possible to slowly move the tab to the left or right and then upon release it would slowly continue moving off the screen. I changed it so that is now takes the velocity into account. You really have to give the tab a push AND be past the treshold. Otherwise it just bounced back.

st3fan commented:
Going to close this one and merge #293
